src/PhaseOne/Artificial.jl

```diff
@@ -64,36 +69,40 @@ function reformulate(lp::Simplex.CanonicalLpData)
         end
     end
 
-    # copy
-    if lp.TArray == Array
-        lp.x = x
-    elseif lp.TArray == CuArray
-        copyto!(lp.x, x)
-    end
-
     # rhs
-    b = Array{Float64}(undef, lp.nrows)
-    copyto!(b, lp.b .- lp.A * lp.x)
+    b = VT(undef, nrows(lp))
+    copyto!(b, lp.b .- lp.A * x)
 
     # create the submatrix for slack
-    I = collect(1:nrows)
-    V = Array{Float64}(undef, nrows)
-    for i = 1:nrows
+    I = collect(1:lp_nrows)
+    V = VT(undef, lp_nrows)
+    for i = 1:lp_nrows
         if b[i] < 0
             V[i] = -1
         else
             V[i] = 1
         end
     end
-    S = sparse(I, I, V)
 
-    # create the matrix in canonical form
-    A = [sparse(Matrix(lp.A)) S]
+    if VT <: Array
+        S = sparse(I, I, V)
+        # create the matrix in canonical form
+        A = [sparse(Matrix(lp.A)) S]
+    elseif VT <: CuArray
+        # on GPU
+        A = [lp.A diagm(V)]
```

Comment:
Here, we are building a dense matrix when we are using the GPU, whereas we were using a sparse matrix previously. We should check that doesn't incur a slowdown.